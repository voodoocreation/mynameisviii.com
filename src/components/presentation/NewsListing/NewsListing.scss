@import "../../../scss/variables", "../../../scss/mixins";

.NewsListing {
  padding-bottom: $spacing;

  &:last-child {
    padding-bottom: 0;
  }

  header {
    @include hide-visually();
  }

  a {
    display: block;
    position: relative;
    color: $text-color;
    opacity: 0.8;
    transition: opacity 200ms ease-out;
    overflow: hidden;

    &:hover {
      opacity: 1;

      p {
        transform: translate3d(0, -100%, 0);
      }
    }
  }

  &--image {
    position: relative;
    width: 100%;
    overflow: hidden;

    .Loader {
      padding-top: 31.25%;
    }
  }

  @include when-below($screen-xs-max) {
    p {
      @include hide-visually();
    }
  }

  @include when-above($screen-sm-min) {
    header {
      position: absolute;
      top: 0;
      left: 0;
      margin-right: $spacing-double;

      &::before {
        @include panel-2();
        @include angle-after-top();

        width: $spacing-double;
      }
    }

    p {
      @include panel-2();

      position: absolute;
      top: 100%;
      right: 0;
      margin: 0 0 0 $spacing-double;
      padding: $spacing-half $spacing;
      transform: translate3d(0, 0, 0);
      transition: transform 200ms ease-out;
      z-index: 1;

      span {
        display: inline-block;
        max-height: 83px;
        overflow: hidden;
      }

      &::before {
        @include panel-2();
        @include angle-before();

        width: $spacing-double;
      }
    }
  }

  @include when-above($screen-md-min) {
    h2,
    h3 {
      font-size: 24px;
    }
  }
}
