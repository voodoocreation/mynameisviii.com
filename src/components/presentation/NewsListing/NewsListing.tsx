import cn from "classnames";
import * as React from "react";

import Schema from "../../schema/NewsArticle";
import Link from "../Link/Link";
import Type from "../Type/Type";

interface IState {
  isRendered: boolean;
}

interface IProps extends INewsArticle {
  isCondensed?: boolean;
  onLoad?: (slug: string) => void;
}

export default class NewsListing extends React.Component<IProps, IState> {
  public static defaultProps = {
    isCondensed: false
  };

  public readonly state = {
    isRendered: false
  };

  private imageRef: React.RefObject<HTMLImageElement> = React.createRef();

  public componentDidMount() {
    if (this.imageRef.current && this.imageRef.current.complete) {
      setTimeout(this.onLoad, 1);
    }
  }

  public render() {
    const { isCondensed, onLoad, ...article } = this.props;

    const header = isCondensed ? (
      <h3>
        <Type value={article.type} hasLabel={false} />
        <span>{article.title}</span>
      </h3>
    ) : (
      <h2>
        <Type value={article.type} hasLabel={false} />
        <span>{article.title}</span>
      </h2>
    );

    return (
      <article
        className={cn(
          "NewsListing",
          { isCondensed },
          { isRendered: this.state.isRendered }
        )}
      >
        <Link route={`/news/${article.slug}`}>
          <header className="NewsListing-header">{header}</header>

          <img
            alt={article.title}
            onError={this.onLoad}
            onLoad={this.onLoad}
            ref={this.imageRef}
            src={article.imageUrl}
          />

          <p>
            <span>{article.excerpt}</span>
          </p>
        </Link>

        <Schema {...article} />
      </article>
    );
  }

  private onLoad = () => {
    this.setState({
      isRendered: true
    });

    if (this.props.onLoad) {
      this.props.onLoad(this.props.slug);
    }
  };
}
