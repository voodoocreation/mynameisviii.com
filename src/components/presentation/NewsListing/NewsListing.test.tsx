import { mount, render } from "enzyme";
import * as React from "react";

import NewsListing from "./NewsListing";

const g: any = global;

const setup = (fn: any, fromTestProps?: any) => {
  const props = {
    author: "author",
    content: "<p>content</p>",
    createdAt: "2017-10-10T18:00:00",
    excerpt: "excerpt",
    imageUrl: "imageUrl",
    isActive: true,
    slug: "slug",
    title: "title",
    ...fromTestProps
  };

  return {
    actual: fn(<NewsListing {...props} />),
    props
  };
};

describe("[presentation] <NewsListing />", () => {
  beforeEach(() => {
    Object.defineProperty(g.Image.prototype, "complete", {
      value: false
    });
  });

  it("renders correctly", () => {
    const { actual } = setup(render);
    expect(actual).toMatchSnapshot();
  });

  it("renders correctly when `isCondensed` prop is true", () => {
    const { actual } = setup(render, { isCondensed: true });
    expect(actual).toMatchSnapshot();
  });

  it("updates `isRendered` state after image has loaded", () => {
    const { actual } = setup(mount);

    actual.find("img").simulate("load");
    expect(actual.render()).toMatchSnapshot();
  });

  it("triggers `onLoad` prop after image has loaded", () => {
    const { actual, props } = setup(mount, {
      onLoad: jest.fn()
    });

    actual.find("img").simulate("load");
    expect(props.onLoad).toHaveBeenCalled();
  });

  it("updates `isImageLoaded` state when image has previously been loaded", () => {
    Object.defineProperty(g.Image.prototype, "complete", {
      value: true
    });

    const { actual } = setup(mount);

    actual.find("img").simulate("load");
    expect(actual.render()).toMatchSnapshot();
  });
});
