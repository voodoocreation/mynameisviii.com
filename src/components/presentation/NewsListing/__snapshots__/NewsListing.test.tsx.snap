// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[presentation] <NewsListing /> renders correctly 1`] = `
<article
  class="NewsListing"
>
  <a
    href="/news/slug"
  >
    <header
      class="NewsListing-header"
    >
      <h2>
        <div
          class="Type Type-undefined"
          title="News"
        >
          <svg
            fill="currentColor"
            height="1em"
            preserveAspectRatio="xMidYMid meet"
            style="vertical-align:middle"
            viewBox="0 0 40 40"
            width="1em"
          >
            <g>
              <path
                d="m21.6 21.6v-10h-3.2v10h3.2z m0 6.8v-3.4h-3.2v3.4h3.2z m16.8-8.4l-4.1 4.6 0.5 6.2-6 1.3-3.1 5.3-5.7-2.4-5.7 2.4-3.1-5.2-6-1.4 0.5-6.2-4.1-4.6 4.1-4.7-0.5-6.1 6-1.3 3.1-5.3 5.7 2.4 5.7-2.4 3.1 5.3 6 1.3-0.5 6.2z"
              />
            </g>
          </svg>
        </div>
        <span>
          title
        </span>
      </h2>
    </header>
    <img
      alt="title"
      src="imageUrl"
    />
    <p>
      <span>
        excerpt
      </span>
    </p>
  </a>
</article>
`;

exports[`[presentation] <NewsListing /> renders correctly when \`isCondensed\` prop is true 1`] = `
<article
  class="NewsListing isCondensed"
>
  <a
    href="/news/slug"
  >
    <header
      class="NewsListing-header"
    >
      <h3>
        <div
          class="Type Type-undefined"
          title="News"
        >
          <svg
            fill="currentColor"
            height="1em"
            preserveAspectRatio="xMidYMid meet"
            style="vertical-align:middle"
            viewBox="0 0 40 40"
            width="1em"
          >
            <g>
              <path
                d="m21.6 21.6v-10h-3.2v10h3.2z m0 6.8v-3.4h-3.2v3.4h3.2z m16.8-8.4l-4.1 4.6 0.5 6.2-6 1.3-3.1 5.3-5.7-2.4-5.7 2.4-3.1-5.2-6-1.4 0.5-6.2-4.1-4.6 4.1-4.7-0.5-6.1 6-1.3 3.1-5.3 5.7 2.4 5.7-2.4 3.1 5.3 6 1.3-0.5 6.2z"
              />
            </g>
          </svg>
        </div>
        <span>
          title
        </span>
      </h3>
    </header>
    <img
      alt="title"
      src="imageUrl"
    />
    <p>
      <span>
        excerpt
      </span>
    </p>
  </a>
</article>
`;

exports[`[presentation] <NewsListing /> updates \`isImageLoaded\` state when image has previously been loaded 1`] = `
<article
  class="NewsListing isRendered"
>
  <a
    href="/news/slug"
  >
    <header
      class="NewsListing-header"
    >
      <h2>
        <div
          class="Type Type-undefined"
          title="News"
        >
          <svg
            fill="currentColor"
            height="1em"
            preserveAspectRatio="xMidYMid meet"
            style="vertical-align: middle;"
            viewBox="0 0 40 40"
            width="1em"
          >
            <g>
              <path
                d="m21.6 21.6v-10h-3.2v10h3.2z m0 6.8v-3.4h-3.2v3.4h3.2z m16.8-8.4l-4.1 4.6 0.5 6.2-6 1.3-3.1 5.3-5.7-2.4-5.7 2.4-3.1-5.2-6-1.4 0.5-6.2-4.1-4.6 4.1-4.7-0.5-6.1 6-1.3 3.1-5.3 5.7 2.4 5.7-2.4 3.1 5.3 6 1.3-0.5 6.2z"
              />
            </g>
          </svg>
        </div>
        <span>
          title
        </span>
      </h2>
    </header>
    <img
      alt="title"
      src="imageUrl"
    />
    <p>
      <span>
        excerpt
      </span>
    </p>
  </a>
</article>
`;

exports[`[presentation] <NewsListing /> updates \`isRendered\` state after image has loaded 1`] = `
<article
  class="NewsListing isRendered"
>
  <a
    href="/news/slug"
  >
    <header
      class="NewsListing-header"
    >
      <h2>
        <div
          class="Type Type-undefined"
          title="News"
        >
          <svg
            fill="currentColor"
            height="1em"
            preserveAspectRatio="xMidYMid meet"
            style="vertical-align: middle;"
            viewBox="0 0 40 40"
            width="1em"
          >
            <g>
              <path
                d="m21.6 21.6v-10h-3.2v10h3.2z m0 6.8v-3.4h-3.2v3.4h3.2z m16.8-8.4l-4.1 4.6 0.5 6.2-6 1.3-3.1 5.3-5.7-2.4-5.7 2.4-3.1-5.2-6-1.4 0.5-6.2-4.1-4.6 4.1-4.7-0.5-6.1 6-1.3 3.1-5.3 5.7 2.4 5.7-2.4 3.1 5.3 6 1.3-0.5 6.2z"
              />
            </g>
          </svg>
        </div>
        <span>
          title
        </span>
      </h2>
    </header>
    <img
      alt="title"
      src="imageUrl"
    />
    <p>
      <span>
        excerpt
      </span>
    </p>
  </a>
</article>
`;
